## step 1 + 2

get data set ready for use

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
df=pd.read_csv(r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\week3\day5\bank.csv")
df.columns
X = df.drop("deposit", axis=1)

y = df["deposit"]

y = y.map({
    "yes": 1,
    "no": 0
})


In [2]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    
)

In [3]:
catcols=X_train.select_dtypes(include='object').columns
numcols=X_train.select_dtypes(exclude='object').columns

C:\Users\Hp\AppData\Local\Temp\ipykernel_17380\4131174849.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  catcols=X_train.select_dtypes(include='object').columns


In [4]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
preprocceser=ColumnTransformer(
    transformers=[
        ('num',StandardScaler(),numcols),
        ('cat',OneHotEncoder(handle_unknown='ignore'),catcols)
    ]
    
)

In [5]:
X_trainpre=preprocceser.fit_transform(X_train)
X_testpre=preprocceser.transform(X_test)


bring the same model

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
model=DecisionTreeClassifier(max_depth=10,random_state=42)

use for loop to see the diffrances and it include 5 folds

In [9]:
num=[2,3,4,5,6,7,8,9,10]
from sklearn.model_selection import cross_val_score
for n in num: 
    scores=cross_val_score(model,X_trainpre,y_train,cv=n,scoring='accuracy')
    print(f'the socore of {n} number of folds the accuracy AVG is: {scores.mean()*100} and the std is: {scores.std()}\n')

the socore of 2 number of folds the accuracy AVG is: 80.94977061734639 and the std is: 0.005018422858794225

the socore of 3 number of folds the accuracy AVG is: 81.75595485339817 and the std is: 0.007674751775248841

the socore of 4 number of folds the accuracy AVG is: 81.96875697223305 and the std is: 0.006691839574982938

the socore of 5 number of folds the accuracy AVG is: 82.24889507874819 and the std is: 0.004654958985399483

the socore of 6 number of folds the accuracy AVG is: 81.7334286560223 and the std is: 0.013792976727443753

the socore of 7 number of folds the accuracy AVG is: 81.84576275651327 and the std is: 0.012478687326261317

the socore of 8 number of folds the accuracy AVG is: 82.65182436313346 and the std is: 0.009786554174810221

the socore of 9 number of folds the accuracy AVG is: 81.71101157926272 and the std is: 0.012580428843470999

the socore of 10 number of folds the accuracy AVG is: 82.41681689674046 and the std is: 0.00879427744720095



## step 3

#### Comparison with Day 1

The single-split accuracy from Day 1 was **80.70%**.

Using 5-fold cross-validation, the mean accuracy was **82.25%** with a standard deviation of **0.00465**.

The cross-validation result is slightly higher than the single-split result. This difference occurs because Day 1 evaluated the model on only one train/valid/test split, while cross-validation evaluates the model across five different folds.

The low standard deviation indicates that the model's performance was relatively stable across the five folds.

Therefore, cross-validation provides a more reliable estimate of the model's performance than relying on a single validation split.


## Step 4 
#### Stratified K-Fold

Since this is a classification problem, `cross_val_score` automatically uses **StratifiedKFold** when `cv=5` is specified for a classifier.

This means that the class proportions are approximately preserved in each fold.

This is important because it ensures that each fold has a similar distribution of classes. It makes the cross-validation results more reliable, especially when the classification dataset is imbalanced.

Therefore, the existing `cv=5` implementation already uses stratified folds, so no code changes are required.
